cpp/include/raft/sparse/matrix/detail/preprocessing.cuh

```diff
@@ -0,0 +1,463 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <raft/core/device_coo_matrix.hpp>
+#include <raft/core/device_mdarray.hpp>
+#include <raft/core/device_span.hpp>
+#include <raft/core/host_mdarray.hpp>
+#include <raft/core/resource/cuda_stream.hpp>
+#include <raft/core/resource/thrust_policy.hpp>
+#include <raft/core/resources.hpp>
+#include <raft/sparse/convert/coo.cuh>
+#include <raft/sparse/op/sort.cuh>
+
+#include <thrust/fill.h>
+#include <thrust/reduce.h>
+
+namespace raft::sparse::matrix::detail {
+
+/**
+ * @brief Calculates the BM25 values for a target matrix. Term frequency is calculate using
+ * logrithmically scaled frequency.
+ * @param num_feats: The total number of features in the matrix
+ * @param avg_feat_len: The avg length of all features combined.
+ * @param k_param: K value required by BM25 algorithm.
+ * @param b_param: B value required by BM25 algorithm.
+ */
+struct bm25 {
+  bm25(int num_feats, float avg_feat_len, float k_param, float b_param)
```

Comment:
you probably want to template this. Maybe an `int_t` and `value_t`. You never think it's going to cause an issue until  you realize someone has num_feats > 1^32-1.

Reply:
Are this and the tfidf structs being used anywhere? I don't see them used.

Reply:
Yes they are used in the base_encode_xxx functions in the final map calls.